src/config.rs

```diff
@@ -0,0 +1,130 @@
+#[derive(Debug, Clone)]
+pub struct Socks5Config {
+    /// The address of the socks5 service
+    pub addr: String,
+    /// Optional credential for the service
+    pub credentials: Option<Socks5Credential>,
+}
+
+/// Credential for the proxy
+#[derive(Debug, Clone)]
+pub struct Socks5Credential {
+    pub username: String,
+    pub password: String,
+}
+
+/// [Config] Builder
+pub struct ConfigBuilder {
+    config: Config,
+}
+
+impl ConfigBuilder {
+    /// Create a builder with a default config, equivalent to [ConfigBuilder::default()]
+    pub fn new() -> Self {
+        ConfigBuilder {
+            config: Config::default(),
+        }
+    }
+
+    /// Set the socks5 config if Some, it accept an `Option` because it's easier for the caller to use
+    /// in a method chain
+    pub fn socks5(mut self, socks5_config: Option<Socks5Config>) -> Result<Self, Error> {
+        if socks5_config.is_some() && self.config.timeout.is_some() {
+            return Err(Error::BothSocksAndTimeout);
+        }
+        self.config.socks5 = socks5_config;
+        Ok(self)
+    }
+
+    /// Sets the timeout
+    pub fn timeout(mut self, timeout: u8) -> Result<Self, Error> {
+        if self.config.socks5.is_some() {
+            return Err(Error::BothSocksAndTimeout);
+        }
+        self.config.timeout = Some(Duration::from_secs(timeout as u64));
+        Ok(self)
+    }
+
+    /// Sets the retry attempts number
+    pub fn retry(mut self, retry: u8) -> Self {
+        self.config.retry = retry;
+        self
+    }
+
+    /// Sets if the domain has to be validated
+    pub fn validate_domain(mut self, validate_domain: bool) -> Self {
+        self.config.validate_domain = validate_domain;
+        self
+    }
+
+    /// Return the config and consume the builder
+    pub fn build(self) -> Config {
+        self.config
+    }
+}
+
+impl Default for ConfigBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Socks5Config {
+    /// Socks5Config constructor without credentials
+    pub fn new(addr: impl ToString) -> Self {
+        let addr = addr.to_string().replacen("socks5://", "", 1);
+        Socks5Config {
+            addr,
+            credentials: None,
+        }
+    }
+
+    /// Socks5Config constructor if we have credentials
+    pub fn with_credentials(addr: impl ToString, username: String, password: String) -> Self {
+        let mut config = Socks5Config::new(addr);
+        config.credentials = Some(Socks5Credential { username, password });
+        config
+    }
+}
+
+impl Config {
+    pub fn socks5(&self) -> &Option<Socks5Config> {
+        &self.socks5
+    }
+    pub fn retry(&self) -> u8 {
+        self.retry
+    }
+    pub fn timeout(&self) -> Option<Duration> {
+        self.timeout
+    }
+    pub fn validate_domain(&self) -> bool {
+        self.validate_domain
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Config {
+            socks5: None,
+            timeout: None,
```

Comment:
Do you know what's the default timeout? Is it reasonable or maybe is it better to override it with something else by default?

Reply:
at the moment it's not convenient in case of proxy to have a default value, because you would need to set it to None before setting the proxy

Reply:
Right, makes sense